base/compiler/ssair/passes.jl

```diff
@@ -548,11 +548,12 @@ function getfield_elim_pass!(ir::IRCode)
             # Also, we should probably have a version of typeassert
             # that's defined not to return its value to make life easier
             # for the backend.
-            pi = insert_node_here!(compact,
-                PiNode(stmt.args[2], compact.result[idx][:type]),
-                compact.result[idx][:type],
-                compact.result[idx][:line], true)
-            compact.ssa_rename[compact.idx-1] = pi
+            Y = stmt.args[2]
+            if !isa(Y, GlobalRef)
```

Comment:
Why doesn't `GlobalRef` work here? Can you explain that in a comment or in the commit message?

Reply:
GlobalRef is not pure, so it is invalid to duplicate it

Reply:
Wasn't the whole point of https://github.com/JuliaLang/julia/pull/36450 to change that? What non-purity remains?

Reply:
They can also still be non-constant

Reply:
If there is a data race you mean? In that case, what if instead of not doing the canonicalization, we did:
```
%a = GlobalRef(...)
typassert(%a, T)
PiNode(%a, T)
```

Reply:
That's fine too